README.md
pyfaceb
=======

Introduction
------------

This python library was started because most of the existing libraries
out there (pyfacebook, facepy, facebook python-sdk) are either no longer
in development, use the old REST API, or are lacking in support of the
more advanced features available, such as batch querying and FQL
support.

Purpose
-------

A lightweight Python wrapper for the Facebook Graph API and Facebook
Query Language (FQL) interface.

Supports basic and advanced features of the Graph API, including batched
API queries, optionally with multiple access_token's.

Basic Usage Examples
--------------------

    fbg = FBGraph('access-token')
     
    me = fbg.get('me')
     
    # prints your first name
    print me['first_name']
     
    # prints out the name of all your likes
    my_likes = fbg.get('me', 'likes')
    for like in my_likes['data']:
        print like['name']
     
    # make a request with parameters...
    my_first_3_likes = fbg.get('me', 'likes', {'limit': 3})
     
    # get an object by it's ID #:
    facebook_platform_page = fbg.get('19292868552')

Batched Query Examples
----------------------

    # batched queries with a single access token
    fbg = FBGraph('access-token')

    # you can write it manually if you like...
    results = fbg.get_batch([
        {'method': 'GET', 'relative_url': 'me'},
        {'method': 'GET', 'relative_url': 'me/friends?limit=50'}
    ])

    # or use the GetRequestFactory
    batch = [GetRequestFactory('me'), GetRequestFactory('me/friends', limit=50)]
    fbg.get_batch(batch)

    # If you need to use multiple access tokens in
    # one batch call, you can!
    # Note, you still must specify a "fallback" token when you create an
    # instance of FBGraph()
    fbg.get_batch([
        GetRequestFactory('PAGE1_FB_OBJ_ID', access_token=PAGE1_ACCESS_TOKEN),
        GetRequestFactory('PAGE2_FB_OBJ_ID', access_token=PAGE2_ACCESS_TOKEN),
        GetRequestFactory('PAGE3_FB_OBJ_ID', access_token=PAGE3_ACCESS_TOKEN),
    ])
   

   




CHANGES.txt
v0.2.4, Jun 18 2013 -- Updated to mock 1.0.1, requests 1.1.0
v0.2.3, Aug 09 2012 -- Disabled SSL verification by default, allow control via constructor.
                       Removed FBQuery & tests.
                       Refactor _issue_request, added pre and post request hooks.
v0.2.2, Aug 08 2012 -- Added ability to specify custom timeouts.
v0.2.1, Jul 23 2012 -- FBHTTPException now contains the HTTP code and body of the response.
v0.2.0, Jun 01 2012 -- FBGraph get(), post() signature change to simplify calls & make orthogonal to batch rqsts
v0.1.9, May 31 2012 -- Refactor requests into single private method. Added specific exception types.
v0.1.8, May 02 2012 -- Fixed problem with dependencies during pip installation.
v0.1.7, Apr 30 2012 -- added post requests, renamed get_batch() to batch(), added unittests
v0.1.6, Apr 18 2012 -- Fixed major bug in FBGraph.get_batch() method.
v0.1.5, Apr 18 2012 -- Added logging, handling of JSON decoding errors.
v0.1.3, Feb 27 2012 -- Fixing issues relating to pip packaging.
v0.1.0, Feb 13 2012 -- Initial release.


pyfaceb/test/test_api.py
import unittest
import requests
from pyfaceb import FBGraph
from mock import patch, Mock
from pyfaceb.exceptions import (FBJSONException, FBHTTPException,
                                FBConnectionException)
from requests.exceptions import SSLError

class FBGraphTest(unittest.TestCase):
    
    def test_basic_get_personal_profile(self):
        expected = {
            u'username': u'kevin.r.stanton',
            u'first_name': u'Kevin',
            u'last_name': u'Stanton',
            u'name': u'Kevin Stanton',
            u'locale': u'en_US',
            u'gender': u'male',
            u'id': u'537208670'
        }
        fbg = FBGraph()
        result = fbg.get('kevin.r.stanton')
        self.assertDictEqual(expected, result)
        
    def test_basic_get_company_profile(self):
        fbg = FBGraph()
        result = fbg.get('SproutSocialInc')
        self.assertIsInstance(result, dict)
        self.assertEquals('2009', result['founded'])
        self.assertEquals('Sprout Social', result['name'])
        self.assertEquals('138467959508514', result['id'])
        self.assertEquals('SproutSocialInc', result['username'])

    @patch.object(requests, 'request')
    def test_get_with_fields_arguments(self, request):
        mock_response = Mock()
        mock_response.text = '{"talking_about_count": 309, "likes": 6201, "id": "138467959508514"}'
        mock_response.status_code = 200
        request.return_value = mock_response
        f = FBGraph()
        result = f.get(
            'sproutsocialinc', {'fields': 'likes,talking_about_count'})
        self.assertDictEqual(result, {
            "id": "138467959508514",
            "likes": 6201,
            "talking_about_count": 309
        })

    @patch.object(requests, 'request')
    def test_FBJSONException(self, request):
        mock_response = Mock()
        mock_response.text = 'i am bad json'
        mock_response.status_code = 200
        request.return_value = mock_response

        fbg = FBGraph()
        self.assertRaises(FBJSONException, fbg.get, 'me')

        try:
            fbg.get('me')
        except FBJSONException as e:
            self.assertEquals(
                'No JSON object could be decoded (i am bad json)', e.message)

    @patch.object(requests, 'request')
    def test_FBHTTPException(self, request):
        mock_response = Mock()
        mock_response.text = 'some fb error'
        mock_response.status_code = 400
        request.return_value = mock_response

        fbg = FBGraph()
        self.assertRaises(FBHTTPException, fbg.get, 'me')
        try:
            data = fbg.get('me')
        except FBHTTPException as e:
            self.assertEquals(e.code, mock_response.status_code)
            self.assertEquals(e.body, mock_response.text)
            self.assertEquals(e.message, 'some fb error')
            self.assertEquals(
                e.__str__(),
                'FBHTTPException(400, \'some fb error\')')
            self.assertEquals(
                e.__repr__(),
                'FBHTTPException(400, \'some fb error\')')


    @patch.object(requests, 'request')
    def test_FBConnectionException(self, request):
        def side_effect(*args, **kwargs):
            raise SSLError('The read operation timed out')
        request.side_effect = side_effect

        fbg = FBGraph()
        self.assertRaises(FBConnectionException, fbg.get, 'me')
        self.assertRaises(FBConnectionException, fbg.post,
            'me', payload={'junk': 'data'})

    @patch.object(requests, 'request')
    def test_batch_with_individual_rqst_errors(self, request):
        """
        Ensure that when single requests in the batch contain an
        error, the error json is deserialized properly and returned
        for individual requests.
        """
        request.return_value = Mock()
        request.return_value.status_code = 200
        with open('./pyfaceb/test/test_batch_with_individual_rqst_errors.json', 'rb') as jsonfile:
            request.return_value.text = jsonfile.read()

        batch_requests = [
            {'method': 'GET', 'relative_url': 'kevin.r.stanton'},
            {'method': 'GET', 'relative_url': 'sproutsocialinc'}
        ]

        fbg = FBGraph('mocktoken')
        responses = fbg.batch(batch_requests)

        self.assertTrue(len(responses), 2)
        self.assertEquals(responses[0]['code'], 200)
        self.assertEquals(responses[0]['body']['id'], '537208670')
        self.assertEquals(responses[1]['code'], 400)
        self.assertDictEqual(responses[1]['body']['error'], {
            'code': 190,
            'type': 'OAuthException',
            'error_subcode': 460,
            'message': 'Error validating access token: The session has been invalidated because the user has changed the password.'
        })

    @patch.object(requests, 'request')
    def test_request_hooks(self, request):
        request.return_value = Mock()
        request.return_value.status_code = 200
        request.return_value.text = '{}'

        pre_hook = Mock()
        post_hook = Mock()

        f = FBGraph(pre_hook=pre_hook, post_hook=post_hook)

        result = f.get('thing1')
        self.assertDictEqual(result, {})
        result = f.get('thing2')
        self.assertDictEqual(result, {})
        self.assertEquals(pre_hook.call_count, 2)
        self.assertEquals(post_hook.call_count, 2)



pyfaceb/exceptions.py
class FBException(Exception):
    """
    Generic exception as a convenience to the client, if they don't care
    to handle different exception types differently.
    """
    pass

class FBHTTPException(FBException):
    """
    This exception is raised for HTTP errors encountered, i.e. a non-200 HTTP
    status codes.
    """
    def __init__(self, code, body):
        self.code = code
        self.body = body

        # for backwards compatibility
        self.message = body

    def __repr__(self):
        values = (self.__class__.__name__, self.code, self.body)
        return '%s(%s, \'%s\')' % values

    def __str__(self):
        return self.__repr__()

class FBJSONException(FBException):
    """
    This exception is raised when the JSON response cannot be deserialized
    properly.
    """

class FBConnectionException(FBException):
    """
    This exception is raised for errors encountered at the connection level,
    like timeout errors, SSL errors, etc.
    """
    pass

pyfaceb/api.py
import requests
import json
import time
import logging
from requests.exceptions import SSLError, Timeout

from .exceptions import (FBException, FBHTTPException, FBJSONException,
    FBConnectionException)

BASE_GRAPH_URL = "https://graph.facebook.com"
BATCH_QUERY_LIMIT = 50
TIMEOUT = 60.0
VERIFY_SSL = False

log = logging.getLogger(__name__)

def GetRequestFactory(relative_url, **params):
    ''' Returns a properly formed GET request dictionary. '''

    params['method'] = 'GET'
    params['relative_url'] = relative_url

    return params

#TODO: PUT, DELETE request factories

class FBGraph(object):
    def __init__(self, access_token='', timeout=TIMEOUT,
                 pre_hook=None, post_hook=None):
        """
        :param access_token: A valid Facebook access token.
        :param timeout: Per-request HTTP timeout. Default is 60.0 seconds.
        :param pre_hook: Callable that will run before a request is made.
        :param post_hook: Callable that will run after a request is made.
        :return: FBGraph instance.
        """

        self._access_token = access_token
        self._response_fmt = 'json'
        self._timeout = timeout
        self.pre_hook = pre_hook
        self.post_hook = post_hook

    def get(self, relative_url, params=None):
        """
        Query facebook's graph api at relative_url with
        query string parameters params, where params is a python dict.
        """
        params = params or {}
        params['access_token'] = self._access_token

        data = self._issue_request('get', relative_url, params=params,
            timeout=self._timeout)

        return data
    
    def post(self, relative_url, payload=None):
        """
        Publish to the graph.
        Returns a deserialized python object, see
        `Graph API<https://developers.facebook.com/docs/reference/api/>`_
        
        Note: this method requires a valid access token to work.
        
        Example 1::
           data = fbg.post('me/feed', {'message': 'Hello, Facebook World!'})
           print data
           # {u'id': u'537208670_111222333444555666777'}
        
        Example 2::
        
           new_pic = open('my_pic.png', 'rb')
           data = fbg.post('me/photos', {
              'source': new_pic,
              'message': 'Hey, I\'m posting a picture on Facebook!'})
           print data
           # {u'id': u'123456789012', u'post_id': u'537208670_123456789012'}
        """
        if not isinstance(payload, dict):
            raise FBException('Must specify payload as dict.')

        files = {}
        for k in payload.keys():
            if isinstance(payload[k], file):
                files[k] = payload[k]
                del payload[k]

        payload = payload or {}
        payload['access_token'] = self._access_token
        
        data = self._issue_request('post', relative_url, data=payload,
            files=files, timeout=self._timeout)

        return data

    def batch(self, batch):
        """
        Query's facebook's graph api in batches. batch is a list of dicts,
        where each dict is of the form::
        
           {'method': 'GET', 'relative_url': 'someurl', ...}.
           
        There are optional params available, see:
        https://developers.facebook.com/docs/reference/api/batch/

        Returns a list of completely deserialized python data structures for
        each batch response, of the form:
            [
                {
                    "headers": [{ ... }],
                    "code": 200,
                    "body": { ... }
                },
                ...
            ]

        If there was an error in an individual response, the "body" will
        contain the deserialized Facebook error response.
        See: https://developers.facebook.com/docs/reference/api/batch/
        """
        payload = {
            'batch': json.dumps(batch),
            'access_token': self._access_token
        }

        data = self._issue_request('post', '', data=payload,
            timeout=self._timeout)
        
        # deserialize the body of each batch response, need to make sure it
        # is deserializable, thanks to this bug:
        # https://developers.facebook.com/bugs/295201867209494
        for d in data:
            if isinstance(d, dict) and 'body' in d:
                try:
                    d['body'] = json.loads(d['body'])
                except Exception as e:
                    log.warn("Error decoding JSON in batched request: {0}".format(e.message))
                    d['body'] = {'data': []} # default in case deserialization fails
                    pass
        
        return data

    def _issue_request(self, method, relative_url, **kwargs):
        """
        Generic method for making requests to the Graph API and deserializing
        the response. Here we aggregate all general error-handling & exception
        catching/raising.

        Returns: deserialized JSON as native Python data structures.
        """
        url = BASE_GRAPH_URL + ('/%s' % relative_url)
        kwargs['timeout'] = kwargs.get('timeout', TIMEOUT)
        kwargs['verify'] = kwargs.get('verify', VERIFY_SSL)

        self._exec_hook('pre')
        try:
            r = requests.request(method, url, **kwargs)
        except (SSLError, Timeout) as e:
            raise FBConnectionException(e.message)
        self._exec_hook('post')

        if r.status_code != requests.codes.ok:
            raise FBHTTPException(r.status_code, r.text)

        try:
            data = json.loads(r.text)
        except ValueError as e:
            log.warn("Error decoding JSON: {0}. JSON={1}".format(e.message, r.text))
            raise FBJSONException("%s (%s)" % (e.message, r.text))

        return data

    def _exec_hook(self, hook_type):
        hook = None

        if hook_type == 'pre':
            hook = self.pre_hook
        elif hook_type == 'post':
            hook = self.post_hook

        if callable(hook):
            try:
                hook()
            except Exception as e:
                log.warning('Warning: pre_hook raised %s', e.message)

